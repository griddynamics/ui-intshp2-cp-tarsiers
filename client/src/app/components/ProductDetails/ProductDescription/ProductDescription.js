import React, { Component } from 'react';
import { Flags } from 'react-feature-flags';
import { Notify } from 'react-redux-notify';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import { productType } from '../../../types';
import appConfig from '../../../../config/appConfig';
import NotifyService from '../../../../utils/notify.service';
import { addItem, removeItem } from '../../../../utils/wishlist.service';

import styles from './ProductDescription.module.scss';

const CartButton = props => {
  const { inCart, item, addToCart, createNotification } = props;
  const cartAction = () => {
    if (!inCart) {
      addToCart(item);
      createNotification(NotifyService.cartAdd);
    }
  };

  return (
    <Link
      to={inCart ? '/cart' : `/products/${item._id}`}
      className={styles.btn_cart}
      onClick={cartAction}
      data-type="cart-btn"
    >
      <span className={inCart ? styles.btn_visible : styles.btn_invisible}>
        Go to Cart
      </span>
      <span className={!inCart ? styles.btn_visible : styles.btn_invisible}>
        Add to Cart
      </span>
    </Link>
  );
};

class ProductDescription extends Component {
  static propTypes = {
    item: productType,
    wished: PropTypes.bool,
    inCart: PropTypes.bool
  };

  static defaultProps = { item: null, wished: false, inCart: false };

  addItem = addItem.bind(this);

  removeItem = removeItem.bind(this);

  toggleWishList = (e, item) => {
    const { wished } = this.props;

    e.preventDefault();

    const cb = !wished ? this.addItem : this.removeItem;

    cb(item);
  };

  render() {
    const {
      item,
      wished,
      inCart,
      children,
      addToCart,
      createNotification
    } = this.props;

    if (!item) {
      return null;
    }

    const sizes = item.sizes.map((element, index, array) => (
      <React.Fragment key={element}>
        <span className={styles.size}>{element}</span>
        {index + 1 !== array.length ? <span>-</span> : null}
      </React.Fragment>
    ));
    const swatches = item.colors.map(color => {
      const style = { backgroundColor: `${color}` };

      return (
        <span key={color} className={styles.color} style={style}>
          {color}
        </span>
      );
    });

    return (
      <div className={styles.background}>
        <div className="container">
          <h1 className={styles.titlebig} id="test">
            {item.title}
          </h1>
          <section className={styles.section}>
            <div className={styles.preview}>{children}</div>
            <div className={styles.details}>
              <div className={styles.info}>
                <h1 className={styles.title}>{item.title}</h1>
                <h3 className={styles.subtitle}>
                  {appConfig.productDescription.subheader}
                </h3>
                <p className={styles.text}>
                  {appConfig.productDescription.description}
                </p>
              </div>
              <div className={styles.select}>
                <div className={styles.flex_row}>
                  <p className={styles.choose}>Available sizes</p>
                  <div className={styles.sizes}>
                    <div className={styles.scrollBoxSize}>{sizes}</div>
                  </div>
                </div>
                <div className={styles.flex_row}>
                  <p className={styles.choose}>Available colors</p>
                  <div className={styles.colors}>
                    <div className={styles.scrollBoxColor}>{swatches}</div>
                  </div>
                </div>
                <div className={styles.flex_row}>
                  <p className={styles.choose}>Available quantity</p>
                  <div className={styles.quantity}>
                    <span>{item.quantity}</span>
                  </div>
                </div>
              </div>
              <div className={styles.order}>
                <p className={styles.price}>{`Price: ${item.price}$`}</p>
                <div className={styles.buttons_row}>
                  <button type="button" className={styles.iconButton}>
                    <i className="fas fa-globe" />
                  </button>
                  <Notify position={NotifyService.position.topRight} />
                  <div className={styles.iconButton}>
                    {inCart ? (
                      <i className="fas fa-cart-arrow-down highlighted" />
                    ) : (
                      <i className="fas fa-cart-plus" />
                    )}
                  </div>
                  <Flags authorizedFlags={[appConfig.killswitch.wishlist]}>
                    <button
                      type="button"
                      onClick={e => this.toggleWishList(e, item)}
                      title="Add to wish-list"
                      data-type="wishlist-btn"
                      className={styles.iconButton}
                    >
                      <i
                        className={
                          wished ? 'fas fa-heart highlighted' : 'fas fa-heart'
                        }
                      />
                    </button>
                  </Flags>
                  <CartButton
                    addToCart={addToCart}
                    item={item}
                    inCart={inCart}
                    createNotification={createNotification}
                  />
                </div>
              </div>
            </div>
          </section>
        </div>
      </div>
    );
  }
}

export default ProductDescription;
