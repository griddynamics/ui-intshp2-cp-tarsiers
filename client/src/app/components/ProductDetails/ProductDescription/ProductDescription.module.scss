@import '../../../../styles/mixin';
@import '../../../../styles/variables';

.background {
  background-color: $color-white;
  border-top: 0.5px solid $color-light-2;
  padding-bottom: 65px;

  .titlebig {
    display: none;
    color: $color-main;
    font-family: $main-font-bold;
    font-size: 52px;
    text-transform: uppercase;
    line-height: 45px;
    margin-top: 30px;
    text-align: center;
    margin-bottom: 20px;
  }
}

.section {
  @include d-flex(center);

  p {
    font-size: 18px;
    font-family: $main-font-regular;
    font-weight: 400;
  }

  .flex_row {
    @include d-flex();

    align-items: center;
  }

  .choose {
    margin-right: 10px;
  }

  .price {
    font-weight: bold;
  }
}

.preview {
  width: 60%;
}

.details {
  margin: 0;
  margin-top: 100px;
  width: 85%;

  .info {
    @include d-flex(flex-start, column);

    & > .title {
      color: $color-main;
      font-family: $main-font-bold;
      font-size: 52px;
      text-transform: uppercase;
      line-height: 45px;
    }
    & > .subtitle {
      color: $color-dark;
      font-family: $main-font-regular;
      font-weight: 400;
      font-size: 18px;
      margin: 30px 0 15px 0;
    }
    & > .text {
      color: $color-dark;
      font-family: $main-font-regular;
      font-size: 16px;
      line-height: 30px;
      margin: 15px 0 30px 0;
      overflow-x: scroll;
      height: 95px;

      &::-webkit-scrollbar {
        display: none;
      }
    }
  }

  .select {
    padding: 30px 0;
    border-top: $border-light;
    border-bottom: $border-light;
    flex-wrap: wrap;
    @include d-flex();
    margin: 0 10px;

    .scrollPic {
      width: 15%;
      align-self: center;
      margin: 0 auto;
    }

    .scrollBoxSize {
      @include d-flex();
    }

    .scrollBoxColor {
      @include d-flex(flex-end, row);
    }

    .sizes {
      overflow: auto;

      &::-webkit-scrollbar {
        display: none;
      }
    }

    .colors {
      overflow: auto;

      &::-webkit-scrollbar {
        display: none;
      }
    }

    .color {
      font-size: 0px;
      display: block;
      min-width: 15px;
      min-height: 15px;
      border-radius: 50%;
      margin: 2px;
      margin-bottom: 5px;
    }

    .size {
      color: $color-dark;
      font-size: 18px;
      padding: 0 5px;
      text-decoration: none;
      text-transform: uppercase;
      outline: none;
    }

    .quantity {
      @include d-flex() {
        align-items: baseline;
      }

      span {
        color: $color-main;
        font-size: 18px;
        padding: 0 10px;
      }
    }
  }

  .order {
    padding: 30px 0;
    @include d-flex() {
      align-items: center;
    }

    & > .buttons_row {
      @include d-flex() {
        align-items: center;
      }

      .iconButton {
        border: none;
        background: none;
        padding: 0 10px;
        color: $color-dark;
        font-size: 18px;

        &:hover:not(div) {
          color: $color-main;
        }
      }

      .btn_cart {
        display: block;
        color: $color-white;
        background-color: $color-main;
        border-radius: 10rem;
        padding: 7px 14px;
        font-size: 16px;
        min-width: 135px;
        text-align: center;
        cursor: pointer;
        text-transform: uppercase;
        font-weight: light;
        transition: all 0.3s ease;
        will-change: auto;
        position: relative;

        &:hover {
          background-color: $color-dark;
        }

        & > span {
          display: inline-block;
        }

        span.btn_visible {
          position: static;
          visibility: visible;
          opacity: 1;
        }

        span.btn_invisible {
          position: absolute;
          visibility: hidden;
          opacity: 0;
          left: 0;
          right: 0;
          bottom: -50px;
        }
      }
    }
  }
}

@media screen and (max-width: $media-tablet) {
  .background {
    .titlebig {
      display: flex;
      flex-direction: row;
      justify-content: center;
    }
  }
  .section {
    @include d-flex(center, column);
    align-items: center;
  }

  .details {
    margin-top: 20px;
  }

  .preview {
    width: 100%;
  }

  .title {
    display: none;
  }
}

@media screen and (max-width: $media-mobile) {
  .section {
    @include d-flex(center, column);

    align-items: center;
  }

  .preview {
    width: 100%;
  }

  .details {
    .select {
      @include d-flex(unset, column);

      .flex_row {
        align-items: center;
      }
    }

    .order {
      @include d-flex(unset, column);

      align-items: center;

      .price {
        padding: 25px 0;
      }
    }
  }
}

@media screen and (max-width: $media-smallest) {
  .section {
    @include d-flex(center, column);
    align-items: center;
  }

  .preview {
    width: 100%;
  }

  .title {
    text-align: center;
    font-size: 40px !important;
  }

  .details {
    .select {
      @include d-flex(unset, column);

      .flex_row {
        align-items: center;
      }
    }

    .order {
      @include d-flex(unset, column);

      align-items: center;

      .price {
        padding: 25px 0;
      }
    }
  }
}
