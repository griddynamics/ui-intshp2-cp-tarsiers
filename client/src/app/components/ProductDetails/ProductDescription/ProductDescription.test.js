import React from 'react';
import { shallow, mount } from 'enzyme';
import ProductDescription from './ProductDescription';

const products = require('../../../../mocks/products.json');

describe('<ProductDescription />', () => {
  let props;

  beforeEach(() => {
    props = {
      item: products[2],
      wished: false,
      inCart: false,
      addToCart: jest.fn(),
      removeFromCart: jest.fn(),
      createNotification: jest.fn()
    };
  });

  it('should match its snapshot', () => {
    const wrapper = shallow(<ProductDescription {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should return null when item is falsy', () => {
    const wrapper = shallow(<ProductDescription />);

    expect(wrapper.html()).toBe(null);
  });

  xdescribe('toggleWishList', () => {
    it('should call addItem when wished is false', () => {
      const evt = { preventDefault() {} };
      const wrapper = shallow(<ProductDescription {...props} />);
      const spy = jest.spyOn(wrapper.instance(), 'addItem');
      const btn = wrapper.find('button[data-type="wishlist-btn"]');

      btn.simulate('click', evt);
      expect(wrapper.state().heartDisabled).toBe(true);
      expect(spy).toHaveBeenCalled();
    });

    it('should call removeItem when wished is true', () => {
      const evt = { preventDefault() {} };
      const wrapper = shallow(<ProductDescription {...props} />);
      const spy = jest.spyOn(wrapper.instance(), 'removeItem');
      const btn = wrapper.find('button[data-type="wishlist-btn"]');

      wrapper.setProps({ wished: true });
      wrapper.update();

      btn.simulate('click', evt);
      expect(wrapper.state().heartDisabled).toBe(false);
      expect(spy).toHaveBeenCalled();
    });
  });

  xdescribe('cart', () => {
    it('should call addToCart when inCart is false', () => {
      const id = props.item._id;
      const evt = { preventDefault() {} };
      const wrapper = mount(<ProductDescription {...props} />);
      const btn = wrapper.find('button[data-type="cart-btn"]');

      btn.simulate('click', evt);
      expect(props.addToCart).toHaveBeenCalledWith(id);
    });

    it('should call removeFromCart when inCart is true', () => {
      const id = props.item._id;
      const evt = { preventDefault() {} };
      const wrapper = mount(<ProductDescription {...props} />);
      const btn = wrapper.find('button[data-type="cart-btn"]');

      wrapper.setProps({ inCart: true });
      wrapper.update();

      btn.simulate('click', evt);
      expect(props.removeFromCart).toHaveBeenCalledWith(id);
    });
  });
});
