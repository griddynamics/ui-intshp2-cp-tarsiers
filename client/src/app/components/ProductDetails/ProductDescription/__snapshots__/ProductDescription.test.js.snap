// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDescription /> should match its snapshot 1`] = `
<div
  className="background"
>
  <div
    className="container"
  >
    <h1
      className="titlebig"
      id="test"
    >
      Nike Sports Suit
    </h1>
    <section
      className="section"
    >
      <div
        className="preview"
      />
      <div
        className="details"
      >
        <div
          className="info"
        >
          <h1
            className="title"
          >
            Nike Sports Suit
          </h1>
          <h3
            className="subtitle"
          >
            Lorem Ipsum is simply dummy text
          </h3>
          <p
            className="text"
          >
            Lorem Ipsum is simply dummy text of the printing and typesetting industry Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum is simply dummy text of the printing and typesetting industry Lorem Ipsum is simply dummy text of the printing and typesetting industry.Lorem Ipsum is simply dummy text of the printing and typesetting industry Lorem Ipsum is simply dummy text of the printing and typesetting industry
          </p>
        </div>
        <div
          className="select"
        >
          <div
            className="flex_row"
          >
            <p
              className="choose"
            >
              Available sizes
            </p>
            <div
              className="sizes"
            >
              <div
                className="scrollBoxSize"
              >
                <span
                  className="size"
                >
                  s
                </span>
                <span>
                  -
                </span>
                <span
                  className="size"
                >
                  m
                </span>
                <span>
                  -
                </span>
                <span
                  className="size"
                >
                  l
                </span>
                <span>
                  -
                </span>
                <span
                  className="size"
                >
                  xl
                </span>
              </div>
            </div>
          </div>
          <div
            className="flex_row"
          >
            <p
              className="choose"
            >
              Available colors
            </p>
            <div
              className="colors"
            >
              <div
                className="scrollBoxColor"
              >
                <span
                  className="color"
                  key="#171717"
                  style={
                    Object {
                      "backgroundColor": "#171717",
                    }
                  }
                >
                  #171717
                </span>
              </div>
            </div>
          </div>
          <div
            className="flex_row"
          >
            <p
              className="choose"
            >
              Available quantity
            </p>
            <div
              className="quantity"
            >
              <span>
                12
              </span>
            </div>
          </div>
        </div>
        <div
          className="order"
        >
          <p
            className="price"
          >
            Price: 800$
          </p>
          <div
            className="buttons_row"
          >
            <button
              className="iconButton"
              type="button"
            >
              <i
                className="fas fa-globe"
              />
            </button>
            <Connect(Notify)
              position="TopRight"
            />
            <div
              className="iconButton"
            >
              <i
                className="fas fa-cart-plus"
              />
            </div>
            <Flags
              authorizedFlags={
                Array [
                  "showWishlist",
                ]
              }
              exactFlags={false}
              renderOff={[Function]}
              renderOn={[Function]}
            >
              <button
                className="iconButton"
                data-type="wishlist-btn"
                onClick={[Function]}
                title="Add to wish-list"
                type="button"
              >
                <i
                  className="fas fa-heart"
                />
              </button>
            </Flags>
            <CartButton
              addToCart={[MockFunction]}
              createNotification={[MockFunction]}
              inCart={false}
              item={
                Object {
                  "_id": "5c486043e7179a544941a6b2",
                  "brand": "Nike",
                  "category": "male",
                  "colorUrls": Object {
                    "#171717": "https://res.cloudinary.com/so/image/upload/v1548246979/products/cba0707c66f3b196408f9bf6e92cc460--knt-nike.jpg",
                  },
                  "colors": Array [
                    "#171717",
                  ],
                  "price": 800,
                  "quantity": 12,
                  "sizes": Array [
                    "s",
                    "m",
                    "l",
                    "xl",
                  ],
                  "src": "https://res.cloudinary.com/so/image/upload/v1548246979/products/cba0707c66f3b196408f9bf6e92cc460--knt-nike.jpg",
                  "title": "Nike Sports Suit",
                }
              }
            />
          </div>
        </div>
      </div>
    </section>
  </div>
</div>
`;
